# Module 10: Descriptive Statistics and Outliers
***

## What are statistics?

Statistics deal with collecting informative data, interpreting these data, and drawing conclusions about the data. Statistics are used across multiple disciplines and fuel business and policy decisions.

## Two types of statistics
***
#### Descriptive Statistics
Descriptive statistics use numeric and graphical techniques to summarize the characteristics of a set of data and to identify patterns within the dataset. Descriptive statistics also include graphically summarizing data and presenting the information in a meaningful (easy to interpret) way. 

#### Inferential Statistics
Inferential statistics use samples of data to make generalizations about a larger set of data (or a larger population). Inferential statistics use data to make estimates, decisions, and predictions about patterns and trends identified within a set of data. Inferential statistics are used to make conclusions beyond your immediate set of data. 

## Import Libraries and Data

In [28]:
import pandas as pd
import numpy as np
import scipy.stats as stats ## new library alert! ##

## SciPy is used for scientific and mathematic computing; it provides functions for stats, 
# data manipulation, and data visualization

df = pd.read_excel("axisdata.xlsx")

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 999 entries, 0 to 998
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Fname             999 non-null    object
 1   Lname             999 non-null    object
 2   Position          999 non-null    object
 3   Gender            999 non-null    object
 4   Hours Worked      999 non-null    int64 
 5   SalesTraining     999 non-null    object
 6   Years Experience  999 non-null    int64 
 7   Cars Sold         999 non-null    int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


## Qualitative vs Quantitative Data
***

#### Quantitative Data
Data that are recorded on a numeric scale (i.e. age, weight, salary, number of empty seats on a plane, credit score etc). Quantitative variables are also called <b>numeric</b> variables. 

#### Qualitative Data
Data that cannot be measured on a numeric scale (i.e. eye color, type of car, blood type, race etc). These data can only be classified into one group of categories. Qualitative variables are also called <b>categorical</b> variables. Qualitative data also includes text and string data. 

In [2]:
## Which variables are qualitative and which variables are quantitative?

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

## Describing Qualitative Data
***

Qualitative data is best summarized by frequency tables. 

Frequency tables summarize categorical data into a table showing each category, number of observations that fall within each category, and the relative frequency of each category. 

Relative frequency is the number of specific observations out of the total number of observations, or the percentage of the total sample that fall within a specific group. Both representations of categorical data give insight into how common specific categories are within a set of data. 

<b>Bar charts</b> are ideal for graphically representing qualitative data as they visually show the frequency of observations across various groups. 

#### GRAPHIC SHOULD READ N = 150 !!!

<img src="FreqTable.png">

In [13]:
## Frequencies 

df["Gender"].value_counts()

M    510
F    489
Name: Gender, dtype: int64

In [14]:
## Relative Frequencies 

df["Gender"].value_counts(normalize=True)

M    0.510511
F    0.489489
Name: Gender, dtype: float64

### How to convert to %

The output for the value_counts function is shown as a decimal; to convert to a %, you can follow eithe of the steps below:

In [12]:
## convert to % by multiplying value by 100

male = 0.510511 * 100 

print(male)

51.051100000000005


In [23]:
df["Gender"].value_counts(normalize=True)*100

M    51.051051
F    48.948949
Name: Gender, dtype: float64

In [7]:
## displaying output as a percentage with formatted strings

male = 0.510511 ## you do NOT need to multiply the value by 100 

print(f"{male:.2%}") ## the number before the % symbol represents how many digits after the decimal

51.05%


In [3]:
## Multi-variable Frequencies 

pd.crosstab(df["Gender"], df["SalesTraining"], margins = True)

## margins = True will give your crosstab table a "total" column/row

# pd.crosstab(df["Gender"], df["SalesTraining"], margins=False, normalize=False)

SalesTraining    N    Y  All
Gender                      
F              212  277  489
M              201  309  510
All            413  586  999

# { Exercise 1 }

    1. Import the "pokemon.csv" file; name the dataset 'poke'. Preview the first 5 rows. 
    2. Determine the frequencies of unique pokemon types (variable name: Type 1) in this dataset
    3. Determine the relative frequencies of the unique pokemon types (variable name: Type 1) in this dataset
    4. Create a crosstabs table that shows the frequency of pokemon stage (variable name: Stage) and legendary status (variable name: Legendary). 

In [4]:
poke = pd.read_csv("pokemon.csv")

poke.head()

Num        Name Type 1  Type 2  Total  HP  Attack  Defense  SpAtk  SpDef  \
0    1   Bulbasaur  Grass  Poison    318  45      49       49     65     65   
1    2     Ivysaur  Grass  Poison    405  60      62       63     80     80   
2    3    Venusaur  Grass  Poison    525  80      82       83    100    100   
3    4  Charmander   Fire     NaN    309  39      52       43     60     50   
4    5  Charmeleon   Fire     NaN    405  58      64       58     80     65   

   Speed  Stage  Legendary  
0     45      1      False  
1     60      2      False  
2     80      3      False  
3     65      1      False  
4     80      2      False

In [9]:
poke.describe()

Num      Total          HP      Attack     Defense       SpAtk  \
count  151.000000  151.00000  151.000000  151.000000  151.000000  151.000000   
mean    76.000000  407.07947   64.211921   72.549669   68.225166   67.139073   
std     43.734045   99.74384   28.590117   26.596162   26.916704   28.534199   
min      1.000000  195.00000   10.000000    5.000000    5.000000   15.000000   
25%     38.500000  320.00000   45.000000   51.000000   50.000000   45.000000   
50%     76.000000  405.00000   60.000000   70.000000   65.000000   65.000000   
75%    113.500000  490.00000   80.000000   90.000000   84.000000   87.500000   
max    151.000000  680.00000  250.000000  134.000000  180.000000  154.000000   

            SpDef       Speed       Stage  
count  151.000000  151.000000  151.000000  
mean    66.019868   68.933775    1.582781  
std     24.197926   26.746880    0.676832  
min     20.000000   15.000000    1.000000  
25%     49.000000   46.500000    1.000000  
50%     65.000000   70.000000    1.000000  
75%     80.000000   90.000000    2.000000  
max    125.000000  140.000000    3.000000

In [15]:
poke["Type 1"].value_counts()

Water       28
Normal      22
Poison      14
Grass       12
Fire        12
Bug         12
Electric     9
Rock         9
Psychic      8
Ground       8
Fighting     7
Ghost        3
Dragon       3
Ice          2
Fairy        2
Name: Type 1, dtype: int64

In [27]:
poke["Type 1"].value_counts(normalize = True)*100

Water       18.543046
Normal      14.569536
Poison       9.271523
Fire         7.947020
Bug          7.947020
Grass        7.947020
Rock         5.960265
Electric     5.960265
Psychic      5.298013
Ground       5.298013
Fighting     4.635762
Ghost        1.986755
Dragon       1.986755
Ice          1.324503
Fairy        1.324503
Name: Type 1, dtype: float64

In [28]:
pd.crosstab(poke["Legendary"], poke["Stage"], margins = True, normalize = False)

Stage       1   2   3  All
Legendary                 
False      75  56  16  147
True        4   0   0    4
All        79  56  16  151

## Describing Quantitative Data

There are many more techniques for describing quantitative data compared to qualitative. When summarizing quantitative data, the objective is to get a sense of how the data is clustered and the variation in numeric values. 

***

### Measures of Central Tendency

Measures of Central Tendency are estimates of what a "typical" value/data point is within your set of data. A measure of central tendency is a number around which an entire sample of data is spread. We can describe this central position using several statistics, including the mean (average), the median, and the mode. 

Measures of central tendency can be visualized with a frequency distribution (histogram). A frequency distribution is a description of values and how often each value occurs within your dataset. When you graph a frequency distribution a pattern/shape should give insight into which values occured the most/least. 

<img src="NormalDist.jpg">

***

#### Mean

The mean, which is the most commonly known and used measure of central tendency, is the arithmetic average of a set of numeric values. The mean is calculated by adding all the values together and dividing by the total number of values in the set of values. In specific sets of data, the mean is the central point in the frequency distribution aka the hump. 

In [3]:
df.head(2)

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6

In [26]:
## Calculate the Mean of a column 

df["Hours Worked"].mean()

33.727727727727725

In [32]:
df.describe()

Hours Worked  Years Experience   Cars Sold
count    999.000000        999.000000  999.000000
mean      33.727728          3.026026    3.922923
std        8.223454          1.394709    1.527000
min       20.000000          1.000000    1.000000
25%       27.000000          2.000000    3.000000
50%       34.000000          3.000000    4.000000
75%       41.000000          4.000000    5.000000
max       48.000000          5.000000    7.000000

#### Median 

The median is the value which divides the data into two equal parts when the data is arranged from least to greatest. The median will be the middle value if the number of values is odd, otherwise, it will be the average of the two middle values if the number of values is even. 

In [27]:
## Calculate the Median of a column 

df["Hours Worked"].median()

34.0

#### Mode 

The mode is the value that appears most frequently in a sample of data, or the value that has the highest frequency. It is possible for a set of data to have no mode, and it is possible for a set of data to have multiple modes. 

In [28]:
## Calculate the Mode of a column 

df["Hours Worked"].mode()

0    38
dtype: int64

In [33]:
df.head(20)

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0     Jackie    Jackson  Trainee      F            32             Y   
1       Mary  Patterson  Trainee      F            43             N   
2      Tanya      Adams  Trainee      F            28             Y   
3      Tanya  Henderson  Trainee      F            24             Y   
4     Walter   Franklin  Trainee      M            25             Y   
5   Victoria   Ettienne  Trainee      F            29             Y   
6     Martin   Ettienne  Trainee      M            42             N   
7      Peter   Michaels  Trainee      M            43             N   
8     Denise  Henderson  Trainee      F            38             N   
9     Walter      Moore  Trainee      M            32             Y   
10       Sam   Michaels  Trainee      M            45             Y   
11      Jada   Walters   Trainee      F            24             Y   
12      Jack   Ettienne  Trainee      M            21             Y   
13     Tanya     Carter  Trainee      F            40             Y   
14       Tom    Mulgrew  Trainee      M            29             Y   
15     Betty     Looner  Trainee      F            41             Y   
16      John       Rose  Trainee      M            42             Y   
17     Roger       Sapp  Trainee      M            41             Y   
18     Karen  Samuelson  Trainee      F            41             Y   
19     india   Franklin  Trainee      F            36             Y   

    Years Experience  Cars Sold  
0                  1          3  
1                  1          6  
2                  1          2  
3                  1          5  
4                  1          4  
5                  1          2  
6                  1          6  
7                  1          6  
8                  1          5  
9                  1          3  
10                 1          3  
11                 1          4  
12                 1          2  
13                 1          4  
14                 1          5  
15                 1          6  
16                 1          2  
17                 1          3  
18                 1          2  
19                 1          2

In [34]:
df["Cars Sold"].value_counts()

5    206
4    202
3    194
2    172
6    156
1     44
7     25
Name: Cars Sold, dtype: int64

# { Exercise 2 }

    1. Using the pokemon dataset ('poke'), calculate the mean attack value (Attack) in the dataset.  
    2. Determine the median defense value (Defense). Write a single sentence on what your findings mean. 
    3. Determine the mode of the speed (Speed) column. Write a single sentence on what your findings mean. 

In [5]:
poke["Attack"].mean()

72.54966887417218

In [32]:
poke["Defense"].median()

65.0

In [33]:
poke["Speed"].mode()

0    70
1    90
dtype: int64

### Data Distribution and Measurement Sensitivity
***

### Normally distributed data vs. non-normally distributed data

Several statistics are suited for data that follows a <b>normal distribution</b>. When visualized, normally distributed data lies in a symmetrical pattern (bell-shape) where the majority of the data surrounds the mean with decreasing amounts evenly distributed to the left and the right. 

While this is the ideal, real-world data is not always normally distributed. Non-normal distribution is possible and there are several statistics that can still be used with non-normally distributed data. It is important to determine how your data is distributed before you embark on certain statistics. 

#### Distribution and Measure of Central Tendency

The shape of the data helps to determine which measure of central tendency should be reported. Some statistics are very sensitive to non-normal distributions, while others are more resilient. Three of the most common shapes to learn are symmetric, left-skewed, and right-skewed. <b>Skew</b> is simply a measure of the asymmetry of the distribution. <b>In the below images, while not shown, the values are increasing from left to right.</b> 

#### Symmetric Data
* mean, median, and mode are all the same here
* no skewness is apparent
* the distribution is described as symmetric

<img src="Symmetrical.jpg">

***

#### Left-skewed Data
* mean < median
* long tail on the left

<img src="LeftSkew.jpg">

***

#### Right-skewed Data
* mean > median
* long tail on the right

<img src="RightSkew.jpg">

***

### Considerations for Shape

When your data is skewed in either direction, you should consider the best measure of central tendency to report. The median is best used when you have skewed data. This is because the median is less affected by extreme values. 

If your data is skewed, this could be an indication that your data has extreme high or low values that are contributing to the skewness. Later in this lesson, we will talk about how to handle extreme values. 

### Measures of Variability (or Spread)

Measures of Variability, or spread, describe the amount of dispersion in your data sample. In other words, measures of variability, describe how spread out values are. For example, in a sample of 100 students, the average test score is 75. However, not all students will have gotten a 75 - their scores will be spread out above and below the average, some will be higher and some lower. To describe the spread of our data, we can use several statistics, including range, standard deviation, percentiles, and quartiles.

***

#### Range

The range is the difference between the lowest and highest value in a set of data. For example, if the maximum value is 10 and the minimum value is 2, the range is 8. The larger the range, the larger the difference between the highest and lowest values. 

In [6]:
## Calculate the range of values in a column

hrs_range = df['Hours Worked'].max() - df['Hours Worked'].min()

print(hrs_range)

28


#### Standard Deviation (SD)

The SD is a measurement of the average distance between each data value and the mean. You can consider SD as a measure of how spread out the data is from the mean. A low SD indicates that the data points are clustered around the mean, while a high SD indicates the data points are spread out over a wider range of values. 

<img src="https://d20khd7ddkh5ls.cloudfront.net/high_low_standard_deviation.png">

In [35]:
## Calculate the Standard Deviation of values in a column

print("Mean hours worked by all employees:", df["Hours Worked"].mean())

df["Hours Worked"].std()

Mean hours worked by all employees: 33.727727727727725


8.223453795492198

#### Percentiles

Percentiles give insight into the relation of one data point to other data points within the same dataset. When the data values are arranged from lowest to highest, percentiles represent the position of a value within the list of values. For example, if a value is at the 50th percentile - 50% of the data falls below the value, and 50% fall above the value. If a value is at the 85th percentile - 85% of the data falls below the value, and 15% fall above the value. 

In [59]:
## Calculate percentiles of values in a column
## To determine the value that falls at a specific percentile
## np.percentile(data/column, percentile value)

np.percentile(df['Hours Worked'], 85)

## Determine which value falls at the 85th percentile? 
## 85% of the values in Hours Worked are below this value, 15% are above this value. 

44.0

#### Quartiles 

Quartiles are values that divide a set of data into equal quarters (when the data is ordered from lowest to highest). There are <b>three</b> quartile values that divide the data into four parts (see below). 

<img src="quar.png">

* The first quartile (Q1) is at the 25th percentile (25% below, 75% above)
* The second quartile (Q2) is at the 50th percentile (50% below, 50% above, median)
* The third quartile (Q3) is at the 75th percentile (75% below, 25% above)

<b>Quartiles can be used to calculate the Interquartile Range (IQR) - which can be used to handle outliers in your dataset.</b>

In [37]:
## Determine the three quartiles of a set of data
## A quantile is a series of values, or cut points, that divide a set of data into equal parts 
## A 'quartile' is a specical kind of quantile that divides the data into four equal parts

df['Hours Worked'].quantile(.25) 

#Q1 = .25
#Q2 = .50
#Q3 = .75

27.0

In [39]:
df['Hours Worked'].quantile(.5) 

34.0

In [38]:
df['Hours Worked'].quantile(.75) 

41.0

#### All of the Above 

The describe function is perfect for outputting multiple descriptive statistics at once. This will save you multiple steps if you want to look at one or several quantitative variables. 

In [38]:
## Use the describe function to view multiple statistics at the same time

df.describe()

Hours Worked  Years Experience   Cars Sold
count    999.000000        999.000000  999.000000
mean      33.727728          3.026026    3.922923
std        8.223454          1.394709    1.527000
min       20.000000          1.000000    1.000000
25%       27.000000          2.000000    3.000000
50%       34.000000          3.000000    4.000000
75%       41.000000          4.000000    5.000000
max       48.000000          5.000000    7.000000

In [39]:
## Use the describe function to isolate information for a single column

df['Hours Worked'].describe()

count    999.000000
mean      33.727728
std        8.223454
min       20.000000
25%       27.000000
50%       34.000000
75%       41.000000
max       48.000000
Name: Hours Worked, dtype: float64

# { Exercise 3 }

    1. Using the pokemon dataset ('poke'), calculate the range for the HP column. 
    2. What is the mean and the standard deviation for the special attack (SpAtk) column?
    3. For the Attack column, determine which value falls at the 80th percentile? Which value falls at the 15th percentile? For each of these values, write a single sentence describing what this means. 
    4. Calculate the second quartile (or the median) for the special defense (SpDef) column. Then, calculate the median of the same column using the method shown above. Are the values the same? 

In [8]:
hp_range = poke['HP'].max() - poke['HP'].min()

print(hp_range)

240


In [52]:
poke["SpAtk"].mean()

67.13907284768212

In [57]:
poke["SpAtk"].std()

28.53419930191353

In [61]:
np.percentile(poke['Attack'], 80)

## 80% of the values in Attack are below this value, 20% are above this value.

95.0

In [7]:
poke['SpDef'].quantile(.5)

65.0

In [9]:
poke["SpDef"].median()

65.0

## Handling and Removing Outliers

***

Outliers are data points that differ significantly from other data points. Outliers are an important feature of a dataset that should be identified and addressed because extreme values can influence the results of your statistics. 

   * In a sample of 5 students, their ages are 25, 22, 97, 23, 21 - 97 is an outlier because the value differs significantly from the remaining data.
   * Without removing outlier, the mean is: 37.6 and the median is: 23
   * When removing the outlier, the mean is: 22.75 and the median is: 22.5

Determining which values are extreme can be challenging when you have a larger dataset. There are several methods that can be used to strategically determine which values are outliers, we will cover two separate methods below. 

In [10]:
df = pd.read_csv("SalaryData.csv")
df.head()

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0  Richard              5.1    66029   41  Female             0  Central   
1     Nick              5.9  8136300   39  Female             0  Central   
2   Morgan              9.5   116969   50  Female             1  Central   
3    Susan              3.2    64445   27    Male             0  Central   
4  Matthew              8.2   113812   47  Female             1  Central   

  CompanyCar  Shift  PerformanceReviewScore  
0          Y      2                    98.3  
1          Y      3                   105.7  
2          N      1                   145.0  
3          Y      3                    75.1  
4          N      2                   131.1

## Method 1 : Using z-scores to detect and remove outliers

In statistics, the <font color=salmon><b>Empirical Rule</b></font> estimates how data, following a normal distribution, will spread around the mean. According to this rule, basically all data values will fall within 3 SD's above or below the mean. The Empirical Rule is also known as the <font color=salmon><b>68-95-99.7 Rule</b></font> because according to this rule, your data will typically fall into the following pattern:

* 68% of the data falls within one SD of the mean

* 95% of the data falls within two SD's of the mean

* 99.7% of the data falls within three SD's of the mean

<img src="SD Bell Curve.jpeg">

#### Empirical Rule Example:

The average weight of an American man is 175 lbs with a SD of 10 lbs. According to the Empirical Rule, we can assume: 

* 68% of American men are between 165 and 185 lbs (or +/- 10 lbs from the mean, or 1 SD)
* 95% of American men are between 155 and 195 lbs (or +/- 20 lbs from the mean, or 2 SD's)
* 99.7% of American men are between 145 and 205 lbs (or +/- 30 lbs from the mean, or 3 SD's)
* Typically, data values that fall outside of this range are considered outliers
_______________________ 
A <font color=salmon><b>z-score</b></font>, also known as a standard score, measures the distance of a value from the mean. In other words, the z-score tells you how many SD's a given value is from the mean. Values that fall above the mean will have a positive (+) z-score, while values that fall below the mean will have a negative (-) z-score. Values that are far from the mean will be considered outliers. You can define a value as an outlier if the z-score greater than +/- 3. 
_______________________ 

#### Steps for Detecting and Removing Outliers with z-scores
* Import the scipy library to make use of the z-score function
* Calculate the z-score for each value within a specific column in your dataset
* Transform the z-score to the absolute value of the z-score (number without the +/- sign; this will make our code simpler)
* Add the z-scores to your dataset as a new column
* Remove all rows that have a z-score greater than 3

In [11]:
## Create a copy of your dataset to filter outliers
dfz = df.copy()

## Check original shape of the dataset
print(dfz.shape)

(30, 10)


In [13]:
## Create a new column to contain the z-scores for the employee salary
## Set the new column equal to the absolute value of the z-scores
## To calculate z-score -- stats.zscore(data/column name)
## To determine absolute value of z-score -- np.abs(data/function/options)

dfz["zscore_salary"] = np.abs(stats.zscore(dfz["Salary"]))

## Preview new column; optional
dfz.head(30)

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0   Richard              5.1    66029   41  Female             0  Central   
1      Nick              5.9  8136300   39  Female             0  Central   
2    Morgan              9.5   116969   50  Female             1  Central   
3     Susan              3.2    64445   27    Male             0  Central   
4   Matthew              8.2   113812   47  Female             1  Central   
5   Richard             10.3      160   47    Male             1  Central   
6     James              3.0    60150   43    Male             1     West   
7     Smith              3.9    63218   49  Female             1     West   
8      Bill             10.5   121872   44    Male             1     West   
9   Richard              2.2    39891   40    Male             0  Central   
10    James              5.3    83088   40  Female             0  Central   
11   Morgan              8.7     1094   27  Female             0  Central   
12   Thomas              1.5   377310   46    Male             1  Central   
13  Richard              4.0    55794   48    Male             1  Central   
14   Rachel              1.3    46205   32  Female             0     East   
15    Susan              4.5     6111   50    Male             1  Central   
16     Alex              6.8    91738   31  Female             0  Central   
17  Matthew              1.1    39343   38  Female             0     East   
18     Alex              2.9    56642   37  Female             0     East   
19    Smith              3.7      300   42  Female             0     East   
20   Rachel              4.1    57081   27    Male             0     East   
21    Susan              7.9   101302   29    Male             0     East   
22  Richard              9.6   112635   48  Female             1     East   
23     Bill              3.2    54445   38    Male             1  Central   
24  Matthew              6.0    93940   33    Male             0  Central   
25    Smith              9.0   105582   28  Female             0  Central   
26     Alex              2.0    43525   34    Male             0     East   
27  Richard              4.9    67938   40  Female             0     East   
28     Bill              7.1  9827300   36  Female             0     East   
29    James              4.0    56957   45    Male             1     West   

   CompanyCar  Shift  PerformanceReviewScore  zscore_salary  
0           Y      2                    98.3       0.269844  
1           Y      3                   105.7       3.343557  
2           N      1                   145.0       0.247036  
3           Y      3                    75.1       0.270553  
4           N      2                   131.1       0.248449  
5           N      3                   152.1       0.299336  
6           Y      1                    77.9       0.272476  
7           N      3                    88.7       0.271102  
8           N      3                   153.2       0.244841  
9           Y      1                    69.0       0.281547  
10          Y      2                   100.0       0.262206  
11          N      1                   130.1       0.298918  
12          Y      1                    63.8       0.130470  
13          Y      2                    89.4       0.274427  
14          Y      1                    57.6       0.278720  
15          Y      3                    95.0       0.296672  
16          Y      1                   112.3       0.258333  
17          Y      1                    57.4       0.281792  
18          Y      3                    75.1       0.274047  
19          Y      3                    84.6       0.299274  
20          Y      2                    84.1       0.273850  
21          Y      2                   122.7       0.254051  
22          N      2                   145.4       0.248976  
23          Y      1                    78.4       0.275031  
24          Y      1                   104.9       0.257347  
25 

In [14]:
## Determine the index locations for the rows with zscores that are greater than "3"

z_outliers = dfz.loc[dfz["zscore_salary"] > 3].index

## Preview list of index values
print(z_outliers)

Int64Index([1, 28], dtype='int64')


In [44]:
## What information can we find at these index locations?

dfz.iloc[[1, 28]]

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
1      Nick              5.9  8136300   39  Female             0  Central   
28     Bill              7.1  9827300   36  Female             0     East   

   CompanyCar  Shift  PerformanceReviewScore  zscore_salary  
1           Y      3                   105.7       3.343557  
28          Y      3                   116.8       4.100689

In [15]:
## Drop rows with above index values
dfz = dfz.drop(z_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfz.shape)


(28, 11)


In [16]:
dfz

Employee  YearsExperience  Salary  Age  Gender  RemoteStatus   Region  \
0   Richard              5.1   66029   41  Female             0  Central   
2    Morgan              9.5  116969   50  Female             1  Central   
3     Susan              3.2   64445   27    Male             0  Central   
4   Matthew              8.2  113812   47  Female             1  Central   
5   Richard             10.3     160   47    Male             1  Central   
6     James              3.0   60150   43    Male             1     West   
7     Smith              3.9   63218   49  Female             1     West   
8      Bill             10.5  121872   44    Male             1     West   
9   Richard              2.2   39891   40    Male             0  Central   
10    James              5.3   83088   40  Female             0  Central   
11   Morgan              8.7    1094   27  Female             0  Central   
12   Thomas              1.5  377310   46    Male             1  Central   
13  Richard              4.0   55794   48    Male             1  Central   
14   Rachel              1.3   46205   32  Female             0     East   
15    Susan              4.5    6111   50    Male             1  Central   
16     Alex              6.8   91738   31  Female             0  Central   
17  Matthew              1.1   39343   38  Female             0     East   
18     Alex              2.9   56642   37  Female             0     East   
19    Smith              3.7     300   42  Female             0     East   
20   Rachel              4.1   57081   27    Male             0     East   
21    Susan              7.9  101302   29    Male             0     East   
22  Richard              9.6  112635   48  Female             1     East   
23     Bill              3.2   54445   38    Male             1  Central   
24  Matthew              6.0   93940   33    Male             0  Central   
25    Smith              9.0  105582   28  Female             0  Central   
26     Alex              2.0   43525   34    Male             0     East   
27  Richard              4.9   67938   40  Female             0     East   
29    James              4.0   56957   45    Male             1     West   

   CompanyCar  Shift  PerformanceReviewScore  zscore_salary  
0           Y      2                    98.3       0.269844  
2           N      1                   145.0       0.247036  
3           Y      3                    75.1       0.270553  
4           N      2                   131.1       0.248449  
5           N      3                   152.1       0.299336  
6           Y      1                    77.9       0.272476  
7           N      3                    88.7       0.271102  
8           N      3                   153.2       0.244841  
9           Y      1                    69.0       0.281547  
10          Y      2                   100.0       0.262206  
11          N      1                   130.1       0.298918  
12          Y      1                    63.8       0.130470  
13          Y      2                    89.4       0.274427  
14          Y      1                    57.6       0.278720  
15          Y      3                    95.0       0.296672  
16          Y      1                   112.3       0.258333  
17          Y      1                    57.4       0.281792  
18          Y      3                    75.1       0.274047  
19          Y      3                    84.6       0.299274  
20          Y      2                    84.1       0.273850  
21          Y      2                   122.7       0.254051  
22          N      2                   145.4       0.248976  
23          Y      1                    78.4       0.275031  
24          Y      1                   104.9       0.257347  
25          N      3                   133.4       0.252134  
26          Y      3                    65.2       0.279920  
27          Y      3                    96.0       0.268989  
29          N      2                    88.5       0.273906

## Method 2 : Using the Interquartile Range (IQR) to detect and remove outliers

The <font color=salmon><b>Interquartile Range (IQR)</b></font> is the difference between the first and third quartile in a set of data. The IQR can be considered where the bulk of the data values rest. IQR is calculated by the following equation: <B>IQR = Q3 - Q1</B>
<img src="IQR.png">
_____________________

The IQR can be used to create "fences" or cut-off values that determine which values fall outside of the acceptable data range. For example, we can calculate a lower-limit -- all values falling below this limit will be considered an outlier. The <b>1.5 Rule</b> is a commonly used method for determining the upper and lower cut-off limits. According to this rule, a value is considered an outlier if it falls:

* ... below Q1 - (1.5 x IQR)

* ... above Q3 + (1.5 x IQR)
_____________________

#### 1.5 Rule Example:

A professor is tracking the number of absences for each of their students. They have compiled a list of the number of absences: 

    1, 3, [4], 6, 7, [7], 8, 8, [9], 22, 37 

In this list of data, we can find the following values:

* Q1 = 4
* Q2 (median) = 7
* Q3 = 9

With these values, you can easily calculate the IQR:

* <B>IQR</B> = Q3(9) - Q1(4)
* <B>IQR</B> = 5

Once you have the IQR, you can calculate the upper and lower limits for detecting outliers:

* Upper Limit = Q3(9) + [1.5 * IQR(5)] = <b> 16.5 </b>
* Lower Limit = Q1(4) - [1.5 * IQR(5)] = <b> -3.5 </b>

Values that fall above the upper limit are outliers; values that fall below the lower limit are outliers
_______________________ 

#### Steps for Detecting and Removing Outliers with IQR

* Calculate quartiles
* Calculate IQR
* Determine upper and lower limits
* Drop values less than the lower and/or greater than the upper

In [17]:
## Create a copy of original dataset
dfq = df.copy()

## Check original shape of the dataset
dfq.shape

(30, 10)

In [18]:
## Calculate quartiles
q1 = dfq["Salary"].quantile(.25)
q3 = dfq["Salary"].quantile(.75)

print("Q1:", q1)
print("Q3:", q3)

Q1: 48265.0
Q3: 104512.0


In [19]:
## Calculate the IQR
iqr = q3 - q1

print("IQR:", iqr)

IQR: 56247.0


In [20]:
## Determine outlier fences 
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)


print("Upper Limit:", top)
print("Lower Limit:", bottom)

Upper Limit: 188882.5
Lower Limit: -36105.5


In [21]:
## Determine the index locations for rows that fall outside of outlier fences

iqr_outliers = dfq.loc[(dfq['Salary'] > top) | (dfq['Salary'] < bottom)].index

print("INDEX VALUES:", iqr_outliers)

INDEX VALUES: Int64Index([1, 12, 28], dtype='int64')


In [22]:
## what values can we find at these index locations?

dfq.iloc[[1, 12, 28]]

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
1      Nick              5.9  8136300   39  Female             0  Central   
12   Thomas              1.5   377310   46    Male             1  Central   
28     Bill              7.1  9827300   36  Female             0     East   

   CompanyCar  Shift  PerformanceReviewScore  
1           Y      3                   105.7  
12          Y      1                    63.8  
28          Y      3                   116.8

In [23]:
## Drop rows with above index values
dfq = dfq.drop(iqr_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfq.shape)

(27, 10)


In [25]:
dfq["Salary"].mean()

63713.555555555555

# { Module 10 Homework }

1. Import the "babies.xlsx" dataset. See below for information on the columns:

    * bwt - birth weight of newborn baby
    * gestation	- gestation length (weeks)
    * parity - previously pregnant (0 = no; 1 = yes)
    * age - age of mother
    * height - height of mother (inches)	
    * weight - weight of mother (pounds)
    * smoke - smoking status of mother (0 = nonsmoker; 1 = smoker)

In [30]:
df = pd.read_excel("babies.xlsx")

df.head()

bwt  gestation  parity   age  height  weight  smoke
0  120      284.0       0  27.0    62.0   100.0    0.0
1  113      282.0       0  33.0    64.0   135.0    0.0
2  128      279.0       0  28.0    64.0    77.0    1.0
3  123        NaN       0  36.0    69.0   190.0    0.0
4  108      282.0       0  23.0    67.0   125.0    1.0

2. Preview the first few rows of the dataset. Complete the following checks for your dataset:

    * output the summary information for the dataset, what types of variables make up this dataset?
    * is there any missing data in this dataset?

In [32]:
df.info()

## No missing data in this dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1236 entries, 0 to 1235
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   bwt        1236 non-null   int64  
 1   gestation  1223 non-null   float64
 2   parity     1236 non-null   int64  
 3   age        1234 non-null   float64
 4   height     1214 non-null   float64
 5   weight     1200 non-null   float64
 6   smoke      1226 non-null   float64
dtypes: float64(5), int64(2)
memory usage: 67.7 KB


3. Handle the missing data in the dataset -- there isn't much, so we can drop all the rows that have at least one missing value. How many rows were dropped from the dataset?

In [38]:
df.isnull().sum()

bwt           0
gestation    13
parity        0
age           2
height       22
weight       36
smoke        10
dtype: int64

In [40]:
df.loc[df.isnull().any(axis = 1)]

bwt  gestation  parity   age  height  weight  smoke
3     123        NaN       0  36.0    69.0   190.0    0.0
39    128      282.0       0  31.0    65.0     NaN    0.0
42    138      302.0       0  26.0     NaN     NaN    1.0
85    129      274.0       0  29.0    71.0     NaN    1.0
89    114        NaN       0  24.0    67.0   113.0    1.0
...   ...        ...     ...   ...     ...     ...    ...
1044  117      323.0       0  26.0    62.0     NaN    0.0
1177  146      305.0       0  23.0     NaN     NaN    0.0
1190  104      275.0       0  24.0     NaN     NaN    0.0
1192  124        NaN       1  39.0    65.0   228.0    0.0
1215  152      267.0       0  28.0     NaN   119.0    1.0

[62 rows x 7 columns]

In [42]:
df.dropna(inplace = True)
df

bwt  gestation  parity   age  height  weight  smoke
0     120      284.0       0  27.0    62.0   100.0    0.0
1     113      282.0       0  33.0    64.0   135.0    0.0
2     128      279.0       0  28.0    64.0    77.0    1.0
4     108      282.0       0  23.0    67.0   125.0    1.0
5     136      286.0       0  25.0    62.0    93.0    0.0
...   ...        ...     ...   ...     ...     ...    ...
1231  113      275.0       1  27.0    60.0   100.0    0.0
1232  128      265.0       0  24.0    67.0   120.0    0.0
1233  130      291.0       0  30.0    65.0    67.0    1.0
1234  125      281.0       1  21.0    65.0   110.0    0.0
1235  117      297.0       0  38.0    65.0   129.0    0.0

[1174 rows x 7 columns]

4. There are two qualitative variables in this dataset. What are they? How do you know they are qualitative variables?

In [ ]:
## The two qualitative variables are 'parity' and 'smoke'

## We know these two variaables are qualitative because they do have nonnumeric value. 

5. Replace the values in the qualitative variables with meaningful labels that describe the different groups. Use any method you've learned in previous modules to complete this task. 

In [ ]:
df.fillna


df

## DO NOT DO YET 

6. Before we move forward with any statistics, let's identify and remove any outliers from the dataset. Using the IQR method, search for outliers in the 5 numeric variables. This will take some time and organization, be careful with your code! Make sure you keep track of how many rows/outliers are removed. 

In [49]:
## Create a copy of original dataset
dfq = df.copy()

## Check original shape of the dataset
dfq.shape

(1174, 7)

In [51]:
## Calculate quartiles
q1b = dfq["bwt"].quantile(.25)
q3b = dfq["bwt"].quantile(.75)
q1g = dfq["gestation"].quantile(.25)
q3g = dfq["gestation"].quantile(.75)
q1a = dfq["age"].quantile(.25)
q3a = dfq["age"].quantile(.75)
q1h = dfq["height"].quantile(.25)
q3h = dfq["height"].quantile(.75)
q1w = dfq["weight"].quantile(.25)
q3w = dfq["weight"].quantile(.75)


print("Q1:", q1b)
print("Q3:", q3b)
print("Q1:", q1g)
print("Q3:", q3g)
print("Q1:", q1a)
print("Q3:", q3a)
print("Q1:", q1h)
print("Q3:", q3h)
print("Q1:", q1w)
print("Q3:", q3w)



Q1: 108.0
Q3: 131.0
Q1: 272.0
Q3: 288.0
Q1: 23.0
Q3: 31.0
Q1: 62.0
Q3: 66.0
Q1: 114.0
Q3: 139.75


In [55]:
## Calculate the IQR
iqrb = q3b - q1b
iqrg = q3g - q1g
iqra = q3a - q1a
iqrh = q3h - q1h
iqrw = q3w - q1w


print("iqrb:", iqrb)
print("iqrg:", iqrg)
print("iqra:", iqra)
print("iqrh:", iqrh)
print("iqrw:", iqrw)

iqrb: 23.0
iqrg: 16.0
iqra: 8.0
iqrh: 4.0
iqrw: 25.75


In [58]:
## Determine outlier fences 
topb = q3b + (iqrb * 1.5)
bottomb = q1b - (iqrb * 1.5)
topg = q3g + (iqrg * 1.5)
bottomg = q1g - (iqrg * 1.5)
topa = q3a + (iqra * 1.5)
bottoma = q1a - (iqra * 1.5)
toph = q3h + (iqrh * 1.5)
bottomh = q1h - (iqrh * 1.5)
topw = q3w + (iqrw * 1.5)
bottomw = q1w - (iqrw * 1.5)


print("Upper Limit:", topb)
print("Lower Limit:", bottomb)
print("Upper Limit:", topg)
print("Lower Limit:", bottomg)
print("Upper Limit:", topa)
print("Lower Limit:", bottoma)
print("Upper Limit:", toph)
print("Lower Limit:", bottomh)
print("Upper Limit:", topw)
print("Lower Limit:", bottomw)

Upper Limit: 165.5
Lower Limit: 73.5
Upper Limit: 312.0
Lower Limit: 248.0
Upper Limit: 43.0
Lower Limit: 11.0
Upper Limit: 72.0
Lower Limit: 56.0
Upper Limit: 178.375
Lower Limit: 75.375


In [97]:
## Determine the index locations for rows that fall outside of outlier fences

iqr_outliers = dfq.loc[[[(dfq['bwt'] > top) | (dfq['bwt'] < bottom)]], [[(dfq['gestation'] > top) | (dfq['gestation']] < bottom)]], [[(dfq['age'] > top) | (dfq['age'] < bottom)]], [[(dfq['height'] > top) | [(dfq['height'] < bottom)]], [[(dfq['weight'] > top) | (dfq['weight'] < bottom)]]].index

print("INDEX VALUES:", iqr_outliers)

SyntaxError: closing parenthesis ']' does not match opening parenthesis '(' (<ipython-input-97-7a356dfce86e>, line 3)

7. Describe the characteristics of your qualitative variables by doing the following:

    * Determine the frequencies of each group within each categorical variable. Which groups have the highest frequency?
    * Determine the relative frequencies of each group within each categorical variable. Which groups have the highest relative frequencies?
    * Create a crosstab table using both the categorical variables in your dataset. How many mothers are smokers who have been pregnant before?

In [68]:
df.head(1)

bwt  gestation  parity   age  height  weight  smoke
0  120      284.0       0  27.0    62.0   100.0    0.0

In [69]:
df["parity"].value_counts()

0    866
1    308
Name: parity, dtype: int64

In [70]:
df["smoke"].value_counts()

0.0    715
1.0    459
Name: smoke, dtype: int64

In [72]:
pd.crosstab(df["parity"], df["smoke"])

smoke   0.0  1.0
parity          
0       525  341
1       190  118

In [71]:
pd.crosstab(df["parity"], df["smoke"], normalize = True)

smoke        0.0       1.0
parity                    
0       0.447189  0.290460
1       0.161840  0.100511

8. Describe the Measures of Central Tendency (mean, median, mode) of the numeric variables in the dataset. Calculate the mean, median, and mode for all 5 numeric variables. For each median and mode, describe the following:

    * What does the median value tell you about the data?
    * What does the mode value tell you about the data?

In [83]:

dfbwtm = df["bwt"].mean()
dfgm = df["gestation"].mean()
dfam = df["age"].mean()
dfhm = df["height"].mean()
dfwm = df["weight"].mean()

print(dfbwtm)
print(dfgm)
print(dfam)
print(dfhm)
print(dfwm)

119.46252129471891
279.1013628620102
27.228279386712096
64.04940374787053
131.32623509369677


In [84]:
dfbwtmed = df["bwt"].median()
dfgmed = df["gestation"].median()
dfamed = df["age"].median()
dfhmed = df["height"].median()
dfwmed = df["weight"].median()

print(dfbwtmed)
print(dfgmed)
print(dfamed)
print(dfhmed)
print(dfwmed)

120.0
280.0
26.0
64.0
125.0


In [85]:
dfbwtmo = df["bwt"].mode()
dfgmo = df["gestation"].mode()
dfamo = df["age"].mode()
dfhmo = df["height"].mode()
dfwmo = df["weight"].mode()

print(dfbwtmo)
print(dfgmo)
print(dfamo)
print(dfhmo)
print(dfwmo)


0    115
dtype: int64
0    282.0
dtype: float64
0    23.0
dtype: float64
0    64.0
dtype: float64
0    130.0
dtype: float64


In [89]:
## Check


df["bwt"].mode()

0    115
dtype: int64

9. Describe the Measures of Variability of the numeric variables in the dataset. Calculate the range, standard deviation, and 85th percentile for all 5 numeric variables. Answer the following questions:

    * Which column has the largest range? Which has the smallest? What does this mean?
    * Which column has the largest standard deviation? Which has the smallest? What does this tell you about those variables?

In [90]:
## Range

## Weight column has the largest range while age column has the smallest range

bwt_range = df['bwt'].max() - df['bwt'].min()
gestation_range = df['gestation'].max() - df['gestation'].min()
age_range = df['age'].max() - df['age'].min()
height_range = df['height'].max() - df['height'].min()
weight_range = df['weight'].max() - df['weight'].min()

print(bwt_range)

print(gestation_range)

print(age_range)

print(height_range)

print(weight_range)

121
205.0
30.0
19.0
1473.0


In [93]:
## Which column has the largest standard deviation? Which has the smallest? What does this tell you about those variables?

df.describe()

bwt    gestation       parity          age       height  \
count  1174.000000  1174.000000  1174.000000  1174.000000  1174.000000   
mean    119.462521   279.101363     0.262351    27.228279    64.049404   
std      18.328671    16.010305     0.440100     5.817839     2.526102   
min      55.000000   148.000000     0.000000    15.000000    53.000000   
25%     108.000000   272.000000     0.000000    23.000000    62.000000   
50%     120.000000   280.000000     0.000000    26.000000    64.000000   
75%     131.000000   288.000000     1.000000    31.000000    66.000000   
max     176.000000   353.000000     1.000000    45.000000    72.000000   

            weight        smoke  
count  1174.000000  1174.000000  
mean    131.326235     0.390971  
std      54.103828     0.488176  
min      27.000000     0.000000  
25%     114.000000     0.000000  
50%     125.000000     0.000000  
75%     139.750000     1.000000  
max    1500.000000     1.000000

In [ ]:
## From the 5 quantitative variables, weight has the largest standard deviation while height has the smallest. This tells us that the variability in weight of the people in the given data is very high that would be unproportional to the height.